wallet/wallet.go

```diff
@@ -0,0 +1,522 @@
+	TxnSourceContract         TransactionSource = "contract"
+	TxnSourceFoundationPayout TransactionSource = "foundation"
+)
+
+var (
+	// ErrNotEnoughFunds is returned when there are not enough unspent outputs
+	// to fund a transaction.
+	ErrNotEnoughFunds = errors.New("not enough funds")
+)
+
+type (
+	// A TransactionSource is a string indicating the source of a transaction.
+	TransactionSource string
+
+	// A Transaction is a transaction relevant to a particular wallet, paired
+	// with useful metadata.
+	Transaction struct {
+		ID          types.TransactionID `json:"id"`
+		Index       types.ChainIndex    `json:"index"`
+		Transaction types.Transaction   `json:"transaction"`
+		Inflow      types.Currency      `json:"inflow"`
+		Outflow     types.Currency      `json:"outflow"`
+		Source      TransactionSource   `json:"source"`
+		Timestamp   time.Time           `json:"timestamp"`
+	}
+
+	// A ChainManager manages the current state of the blockchain.
+	ChainManager interface {
+		BestIndex(height uint64) (types.ChainIndex, bool)
+
+		PoolTransactions() []types.Transaction
+
+		AddSubscriber(chain.Subscriber, types.ChainIndex) error
+		RemoveSubscriber(chain.Subscriber)
+	}
+
+	// A SingleAddressStore stores the state of a single-address wallet.
+	// Implementations are assumed to be thread safe.
+	SingleAddressStore interface {
+		chain.Subscriber
+
+		// LastIndexedTip returns the consensus change ID and block height of
+		// the last wallet change.
+		LastIndexedTip() (types.ChainIndex, error)
+		// UnspentSiacoinElements returns a list of all unspent siacoin outputs
+		UnspentSiacoinElements() ([]types.SiacoinElement, error)
+		// Transactions returns a paginated list of transactions ordered by
+		// maturity height, descending. If no more transactions are available,
+		// (nil, nil) should be returned.
+		Transactions(limit, offset int) ([]Transaction, error)
+		// TransactionCount returns the total number of transactions in the
+		// wallet.
+		TransactionCount() (uint64, error)
+		// ResetWallet resets the wallet to its initial state. This is used when a
+		// consensus subscription error occurs.
+		ResetWallet(seedHash types.Hash256) error
+		// VerifyWalletKey checks that the wallet seed matches the existing seed
+		// hash. This detects if the user's recovery phrase has changed and the
+		// wallet needs to rescan.
+		VerifyWalletKey(seedHash types.Hash256) error
+	}
+
+	// A SingleAddressWallet is a hot wallet that manages the outputs controlled by
+	// a single address.
+	SingleAddressWallet struct {
+		priv types.PrivateKey
+		addr types.Address
+
+		cm    ChainManager
+		store SingleAddressStore
+		log   *zap.Logger
+
+		mu  sync.Mutex // protects the following fields
+		tip types.ChainIndex
+		// locked is a set of siacoin output IDs locked by FundTransaction. They
+		// will be released either by calling Release for unused transactions or
+		// being confirmed in a block.
+		locked map[types.Hash256]bool
```

Comment:
would be nice to change this to `locked map[types.Hash256]time.Time` where the timestamp is the last time it was locked like we do in `renterd` and then pass in an argument when creating the wallet to determine for how long an output is locked before trying to spend it again.

It might be different in `hostd` but in `renterd` there is no 100% guarantee that an output is unlocked upon error in a distributed setup.